nvflare/app_common/executors/launcher_executor.py

```diff
@@ -243,6 +221,56 @@ def _init_converter(self, fl_ctx: FLContext):
             check_object_type(self._to_nvflare_converter_id, to_nvflare_converter, ParamsConverter)
             self._to_nvflare_converter = to_nvflare_converter
 
+    def _launch_external_process(
+        self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal
+    ) -> bool:
+        self._abort_signal = abort_signal
+        # if not launched yet
+        if not self._launch_once or not self._launched.is_set():
+            self.prepare_config_for_launch(shareable, fl_ctx)
+            launch_success = self._launch(task_name, shareable, fl_ctx, abort_signal)
+            if not launch_success:
+                self.log_error(fl_ctx, f"launch task ({task_name}): failed")
+                return False
+            self._launched.set()
+            self.log_info(fl_ctx, f"External process for task ({task_name}) is launched.")
+        # wait for external process to set up their pipe_handler
+        setup_success = self._wait_external_setup(task_name, fl_ctx, abort_signal)
+        if not setup_success:
+            self.log_error(fl_ctx, "External process set up failed.")
+            return False
+        return True
+
+    def _end_external_process(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal):
+        if not self._launch_once or self._job_end:
+            ask_peer_end_success = self.ask_peer_to_end(fl_ctx)
+            if not ask_peer_end_success:
+                return False
+            launch_finish = self._wait_launch_finish(task_name, shareable, fl_ctx, abort_signal)
+            if not launch_finish:
+                return False
+            self._clear_state()
+            self.log_info(fl_ctx, f"Launched external process for task ({task_name}) is finished.")
+        return True
+
+    def _wait_external_setup(self, task_name: str, fl_ctx: FLContext, abort_signal: Signal):
+        start_time = time.time()
+        while True:
+            if self._launch_timeout and time.time() - start_time >= self._launch_timeout:
+                self.log_error(fl_ctx, f"External process is not set up within timeout: {self._launch_timeout}")
+                return False
+
+            if abort_signal.triggered:
+                return False
+
+            if self.peer_is_up_or_dead():
+                return True
+
+            if self.launcher.check_run_status(task_name, fl_ctx) != LauncherRunStatus.RUNNING:
+                return False
+
+            time.sleep(0.1)
+
     def _launch(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> bool:
         future = self._thread_pool_executor.submit(self._launch_task, task_name, shareable, fl_ctx, abort_signal)
```

Comment:
Why do you need another thread to do this? You are not doing anything in parallel.

Reply:
I was using this because I want to have the pattern below: `future.result(timeout=xxx)`

I can change this to a while loop and avoid the `ThreadPoolExecutor`

Reply:
After discussion, I remember the original reason I did this is because I assume when users write "Launcher", they could make mistake saying a infinite loop in "launch_task" implementation, that's why we use a different thread to end it.

Let me simplify the code and add more comments.

Reply:
wraps into "_execute_launcher_method_in_thread_executor" for clarity.